src/core/ext/client_config/lb_policy_factory.h

```diff
@@ -47,8 +47,19 @@ struct grpc_lb_policy_factory {
   const grpc_lb_policy_factory_vtable *vtable;
 };
 
+/** A resolved address alongside any LB related information associated with it.
+ * \a user_data, if not NULL, contains opaque data meant to be consumed by the
+ * gRPC LB policy. Note that no all LB policies support \a user_data as input.
+ * Those who don't will simply ignore it and will correspondingly return NULL in
```

Comment:
Good question.  Here are my thoughts.

In general, I agree that crashing is a bad idea.  In fact, I suspect that we over-use asserts in gRPC in general, and we should probably make an effort to stop doing that.  From the application author's perspective, it's extremely poor behavior on the part of a library to crash the whole process instead of returning some indication of an error to the caller.

In general, I also don't think it's a good idea to silently ignore input, since that can cause confusion that may be hard to debug.  I think it's better to indicate a failure somehow, so that the caller immediately knows there's a problem.

So, in principle, my ideal choice would be to somehow return an error to the caller.  However, in this particular case, I'm not sure if that's really the best behavior.  Do we really want to fail to create the LB policy just because the caller passes in user data that we don't know how to propagate?  I would certainly not want to accidentally cause a situation where (e.g.) a resolver provides user data but happens to choose an LB policy that doesn't support user data, thus causing RPCs to fail.

Without more than one specific use-case of how callers might want to use the user data, it's hard to say what the best behavior is for the general case.  For example, if we knew that any caller that passes in user data will not be able to function correctly if the user data is not returned with the pick, then it's probably reasonable to fail at LB policy construction if the policy doesn't support user data, because we know that we won't be able to do what the caller wants.  However, if it's possible that the caller will be able to function in some sort of fall-back way without the user data, then the right thing is probably to proceed anyway.

I think we're going to have to continue evaluating this as we find more use-cases in the future.  For now, I propose that we log an error instead of crashing in LB policy instantiation.  This is far from ideal, but it will prevent us from accidentally crashing if an unexpected LB policy is chosen, and it will provide at least some indication of the problem to a developer trying to implement a new resolver or LB policy.

Hmmm... I just had another thought here that might avoid this problem entirely.  What if we created a map in the grpclb policy that maps the backend address to the LB token.  Then, when the RR policy returns the subchannel, we check to see what backend address the subchannel is for, and we look up the LB token for that address in our map.  That way, we could completely avoid adding the user data to the LB policy API, which would eliminate this entire problem.  What do you think?